src/vmm/src/vstate/vcpu/x86_64.rs

```diff
@@ -193,58 +233,70 @@ impl KvmVcpu {
+    ///
+    /// # Errors
+    ///
+    /// When:
+    /// - [`kvm_ioctls::ioctls::vcpu::VcpuFd::set_cpuid2`] errors.
+    /// - [`vmm_sys_util::fam::FamStructWrapper::push`] errors.
+    /// - [`arch::x86_64::msr::set_msrs`] errors.
+    /// - [`arch::x86_64::regs::setup_regs`] errors.
+    /// - [`arch::x86_64::regs::setup_fpu`] errors.
+    /// - [`arch::x86_64::regs::sregs`] errors.
+    /// - [`arch::x86_64::interrupts::set_lint`] errors
     pub fn configure(
         &mut self,
         guest_mem: &GuestMemoryMmap,
         kernel_start_addr: GuestAddress,
         vcpu_config: &VcpuConfig,
         mut cpuid: CpuId,
-    ) -> Result<()> {
-        let cpuid_vm_spec = VmSpec::new(self.index, vcpu_config.vcpu_count, vcpu_config.smt)
-            .map_err(Error::CpuId)?;
-
-        filter_cpuid(&mut cpuid, &cpuid_vm_spec).map_err(|err| {
-            METRICS.vcpu.filter_cpuid.inc();
-            error!(
-                "Failure in configuring CPUID for vcpu {}: {:?}",
-                self.index, err
-            );
-            Error::CpuId(err)
-        })?;
-
-        match vcpu_config.cpu_template {
-            CpuFeaturesTemplate::T2 => {
-                t2::set_cpuid_entries(&mut cpuid, &cpuid_vm_spec).map_err(Error::CpuId)?
-            }
-            CpuFeaturesTemplate::T2S => {
-                t2s::set_cpuid_entries(&mut cpuid, &cpuid_vm_spec).map_err(Error::CpuId)?
-            }
-            CpuFeaturesTemplate::C3 => {
-                c3::set_cpuid_entries(&mut cpuid, &cpuid_vm_spec).map_err(Error::CpuId)?
-            }
-            CpuFeaturesTemplate::None => {}
+    ) -> std::result::Result<(), KvmVcpuConfigureError> {
+        // 
+        // - It seems `cpuid_vm_spec` was never actually used? just passed as a reference but never
+        //   accessed in the transformers. Is this correct?
+        // - `filter_cpuid` was used to set values for all intel cpu's, since we load a full
+        //   template we don't need this masking.
+        // - Notably when looking at the previous implementations it seems `&cpuid_vm_spec` was
+        //   never used. Is this correct?
+
+        // If a template is specified, get the CPUID template.
+        let template_opt: Option<cpuid::Cpuid> = match vcpu_config.cpu_template {
+            #[cfg(feature = "t2")]
+            CpuFeaturesTemplate::T2 => Some(cpuid::Cpuid::T2),
+            #[cfg(feature = "t2s")]
+            CpuFeaturesTemplate::T2S => Some(cpuid::Cpuid::T2S),
+            #[cfg(feature = "c3")]
+            CpuFeaturesTemplate::C3 => Some(cpuid::Cpuid::C3),
+            CpuFeaturesTemplate::None => None,
+        };
+        // If a template is specified, set our CPUID value to this template.
+        if let Some(template) = template_opt {
+            // *cpuid = cpuid::get_template_cpuid(template);
+            let kvm_cpuid = kvm_bindings::CpuId::from(template);
+            cpuid = kvm_cpuid;
         }
+        // Set CPUID in the KVM
+        self.fd
```

Comment:
Essentially we need to modify this CPUID to contain:
* a custom brand string (this can be part of the template)
* custom APIC IDs based on the number of vCPUs configured
* custom topology based on the number of vCPUs configured and `smt` value

Reply:
I believe this is the same issue as noted https://github.com/firecracker-microvm/firecracker/pull/3076#discussion_r952698904.